### ISSUE TITLE
Rule proposal: `prefer-template`

### ISSUE THREAD

**Issue by shermam** [typescript-eslint/typescript-eslint#3958]
### Description

ESLint already has this [`prefer-template`](https://eslint.org/docs/rules/prefer-template) rule, but it has an important limitation. When you use expressions as operands ESLint doesn't know what type the expressions are and will not report those cases.

The basics of that rule is that it enforces the use of templates instead of using the `+` sign for string concatenation.

So this is not allowed:

```ts
const name = "John";
const str = "Hello, " + name; // <- ESLint will flag this
```

and we should use this instead:

```ts
const name = "John";
const str = `Hello, ${name}`
```

But if we slightly change the first example to:

```ts
const name = "John";
const greeting = "Hello, ";
const str = greeting + name; // <- ESLint will not flag this because it does not infer the types of the operands
```

My suggestion here is to have basically the same rule here with the difference of using the inferred types of the operand expressions in order to determine if we should report it or not.

### Fail

```ts
const name = "John";
const str = "Hello, " + name; // <- This should fail

const greeting = "Hello, ";
const str2 = greeting + name; // <- This should also fail
```

### Pass

```ts
const name = "John";
const greeting = "Hello, ";
const str = `Hello, ${name}`; // <- This should pass
const str2 = `${greeting}${name}`; // <- This should also pass
```

**Comment by bradzacher** [comment]
Thanks for the issue.
This is a very specific, opinionated and stylistic enhancement to the base lint rule.
For that reason I don't think it fits within this general-purpose plugin.

Feel free to build on top of our tooling in a new plugin or locally in your project!
